test/UnitSocketLifecycle.cpp

```diff
@@ -0,0 +1,263 @@
+ */
+
+#include <config.h>
+
+#include <memory>
+#include <string>
+
+#include <HttpRequest.hpp>
+#include <Socket.hpp>
+
+#include <Poco/DOM/DOMParser.h>
+#include <Poco/DOM/Document.h>
+#include <Poco/DOM/Node.h>
+#include <Poco/DOM/NodeFilter.h>
+#include <Poco/DOM/NodeIterator.h>
+#include <Poco/SAX/InputSource.h>
+#include <Poco/URI.h>
+#include <test/lokassert.hpp>
+
+#include <Unit.hpp>
+#include <UserMessages.hpp>
+#include <Util.hpp>
+#include <helpers.hpp>
+#include <vector>
+
+/// Test suite that uses a HTTP session (and not just a socket) directly.
+class UnitSocketLifecycle : public UnitWSD
+{
+    TestResult testOneHTTPSession();
+    TestResult testWebsockNativePingPong();
+    TestResult testWebsockWSDChatPingPong();
+
+public:
+    UnitSocketLifecycle()
+        : UnitWSD("UnitSocketLifecycle")
+    {
+    }
+
+    void invokeWSDTest() override;
+};
+
+UnitBase::TestResult UnitSocketLifecycle::testOneHTTPSession()
+{
+    setTestname(__func__);
+    TST_LOG("Starting Test: " << testname);
+
+    const std::string documentURL = "/favicon.ico";
+
+    // Keep alive socket, avoid forced socket disconnect via dtor
+    TerminatingPoll socketPoller(testname);
+    socketPoller.runOnClientThread();
+
+    const int sockCount = 1;
+    // Reused http session, keep-alive
+    std::vector<std::shared_ptr<http::Session>>
+        sessions; //  = http::Session::create(helpers::getTestServerURI());
+
+    try
+    {
+        for(int sockIdx = 0; sockIdx < sockCount; ++sockIdx) {
+            sessions.push_back( http::Session::create(helpers::getTestServerURI()) );
+            if (sockIdx > 0)
+            {
+                LOK_ASSERT_EQUAL(true, sessions[sockIdx]->isConnected());
+            }
+            TST_LOG("Test: " << testname << "[" << sockIdx << "]: `" << documentURL << "`");
+            http::Request request(documentURL, http::Request::VERB_GET);
+            const std::shared_ptr<const http::Response> response =
+                sessions[sockIdx]->syncRequest(request, socketPoller);
+            TST_LOG("Response: " << response->header().toString());
+            TST_LOG("Response size: " << testname << "[" << sockIdx << "]: `" << documentURL << "`: " << response->header().getContentLength());
+            LOK_ASSERT_EQUAL(http::StatusCode::OK, response->statusCode());
+            LOK_ASSERT_EQUAL(true, sessions[sockIdx]->isConnected());
+            LOK_ASSERT(http::Header::ConnectionToken::None ==
+                       response->header().getConnectionToken());
+            LOK_ASSERT(0 < response->header().getContentLength());
+        }
+    }
+    catch (const Poco::Exception& exc)
+    {
+        LOK_ASSERT_FAIL(exc.displayText());
+    }
+    for (std::shared_ptr<http::Session>& session : sessions)
+    {
+        LOK_ASSERT_EQUAL(true, session->isConnected());
+    }
+    TST_LOG("Test: XXX " << testname << " w/ " << sockCount << " connections, sleep...");
+    std::this_thread::sleep_for(std::chrono::seconds(4));
```

Comment:
Looks like an anti-pattern, especially that we don't check anything afterwards. If something is checked elsewhere, a comment is helpful. Also, the amount of sleep should depend on something that affects the test, so needs to be a variable so we don't have to change this arbitrarily when the test fails or when another parameter changes (at least we shouldn't change it here).

Reply:
See above .. a truly WIP unit test.
When I have timeout configurable .. I will use this to trigger timeouts and then naturally check results.